avm/src/daemon/daemon.rs

```diff
@@ -0,0 +1,28 @@
+use once_cell::sync::OnceCell;
+use tokio::task;
+use tokio::time::{Duration, sleep};
+
+use crate::vm::run::run;
+
+pub static APP_ID: OnceCell<String> = OnceCell::new();
+pub static DEPLOY_KEY: OnceCell<String> = OnceCell::new();
+
+pub async fn start(agz_file: &str, app_id: &str, deploy_key: &str) {
+  let set_app_id = APP_ID.set(app_id.to_string());
+  let set_deploy_key = DEPLOY_KEY.set(deploy_key.to_string());
+  if set_app_id.is_err() || set_deploy_key.is_err() {
+    eprintln!("Failed to globally set app id or deploy key");
+    std::process::exit(1);
+  }
+  let _app_id_global = APP_ID.get().unwrap();
+  let _deploy_key_global = DEPLOY_KEY.get().unwrap();
+  task::spawn(async move {
+    let ten_min_period = Duration::from_secs(60*10);
```

Comment:
Ten minutes?

Reply:
How much time did you have in mind?

Reply:
That was a placeholder and I agree it is too high, but I was still thinking what it should actually be

Reply:
So, if we go with a uniform sampling frequency, then the frequency of the updates should be ~2x the speed we want to target potential upgrades/downgrades, because of [Nyquist sampling theory](https://en.wikipedia.org/wiki/Nyquist_frequency), so if we want deploys to happen in ~1 minute, we should update the status of the cluster once every 30 seconds.

However, the real scaling needs are based on the actual load the servers are under, and that's going to vary from cluster to cluster, and some are going to have "spikier" demand than others. Internally checking the stats and updating when some threshold is reached is probably better. I'd imagine there's 3 thresholds we care about:

1. max update interval -- if we've reached it, force an update and reset the internal counter; this is good for keeping an eye on low-utilization or stable-utilization clusters
2. min update interval -- if we've triggered an update very recently, we should not fire another update for some time; this should be at least the expected time to upgrade/downgrade a cluster, ideally 2x.
3. last update stats ratio threshold -- for all of the stats sent to the server, hold onto them as the most-recently sent stats and compare the current stats versus them. If the ratio between the two has significantly changed (say, 50% up or down) then push the update assuming we're in between the two intervals

This approach is more complicated, but should reduce bandwidth usage while also being more responsive at the same time.

Reply:
So something like `max = 30, min = 2, ratio = 0.5` would be what I'd propose as our first pass for the more complicated update logic.

Reply:
That doesn't change my note about making sure that the actual upgrade/downgrade rules make sure to avoid "resonance" causing constant creates/deletes of VMs; this is just about the relative change of the stats, not the absolute values of them.

Reply:
though that does imply a fourth metric to push an update on: absolute threshold. In case the load balancing logic somehow failed, the AVM needs to put out an SOS kind of message, since 50% more than 90% utilization is over 100% utilization, and nonsensical.

Reply:
> So something like `max = 30, min = 2, ratio = 0.5` would be what I'd propose as our first pass for the more complicated update logic.

What are these units?

Reply:
All that sounds good and we can go really deep here, but we have to be careful not to over-engineer for the very first pass. As you put it, perfect is the enemy of good.

Reply:
That's why I put the simple sampling approach first ;)

But the units for the first two are minutes and ratio is unitless.

Reply:
> > So something like `max = 30, min = 2, ratio = 0.5` would be what I'd propose as our first pass for the more complicated update logic.
> 
> What are these units?

Minutes, I believe. Hmm

Reply:
> All that sounds good and we can go really deep here, but we have to be careful not to over-engineer for the very first pass. As you put it, perfect is the enemy of good.

Actually, this might not be too bad to implement initially. I thought you meant storing stats and looking at the derivative, but you just mean keeping the previous value around.

Reply:
> That's why I put the simple sampling approach first ;)
> 
> But the units for the first two are minutes and ratio is unitless.

Oh I think I'm tired. I missed the fact that you were proposing 2 different ones 😴

Reply:
No problem :)

I think a ~1 minute tolerance to increased latencies will be fine for most users, so a 30 second update interval will work, but you can see how you get 60 hits per cluster per minute that way (or 3600 hits per cluster per hour), while the more advanced one *should* have the same sort of reaction speed but potentially only 2 hits per cluster per *hour* and a maximum of only 30 hits per cluster per hour, so literally 120x - 1800x less traffic for the same number of managed clusters.

Reply:
Changing to 30 seconds and created https://github.com/alantech/deploy/issues/14 as a followup 🚀